dslog.8
.\" Man page generated from reStructuredText.
.
.TH DLOG 8 "" "" ""
.SH NAME
dlog \- Send lines from standard input to the system log
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBdslog [options] name\fP
.SH DESCRIPTION
.sp
The \fBdslog\fP program sends lines given as standard input to the system
logger, one line at a time, with a selectable priority, facility and origin
program name.
.SH USAGE
.sp
Command line options:
.INDENT 0.0
.TP
.BI \-p \ PRIORITY\fR,\fB \ \-\-priority \ PRIORITY
Priority of messages. Refer to \fIsyslog(3)\fP to see possible
values. Just pass any valid priority without the \fBLOG_\fP
prefix. Case does not matter.
.TP
.BI \-f \ FACILITY\fR,\fB \ \-\-facility \ FACILITY
Logging facility. Refer to \fIsyslog(3)\fP to see possible values.
Just pass any valid facility without the \fBLOG_\fP prefix. Case
does not matter.
.TP
.BI \-i \ NUMBER\fR,\fB \ \-\-input\-fd \ NUMBER
Use file descriptor \fBNUMBER\fP to read input. By default the
standard input descriptor (number \fB0\fP) is used.
.TP
.B \-c\fP,\fB  \-\-console
If a message cannot be sent to the system logger, print a copy
of it to the system console.
.TP
.B \-e\fP,\fB  \-\-skip\-empty
Ignore empty input lines. An empty line is one that does not
contain any characters; a line which contains whitespace is
\fBnot\fP considered empty.
.TP
.B \-h\fP,\fB  \-\-help
Show a summary of available options.
.UNINDENT
.sp
Albeit it can be used stan\-alone, most of the time you will be running
\fBdslog\fP under a process control tool like \fIdmon(8)\fP or \fIsupervise(8)\fP\&.
.SH ENVIRONMENT
.sp
Additional options will be picked from the \fBDSLOG_OPTIONS\fP environment
variable, if defined. Any command line option can be specified this way.
Arguments read from the environment variable will be prepended to the ones
given in the command line, so they may still be overriden.
.SH SEE ALSO
.sp
\fIdmon(8)\fP, \fIdlog(8)\fP, \fIrotlog(8)\fP, \fImultilog(8)\fP, \fIsupervise(8)\fP
.sp
\fI\%http://cr.yp.to/daemontools.html\fP
.SH AUTHOR
Adrian Perez <aperez@igalia.com>
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.5.1] - 2021-02-23
### Added
- The `dlog`, `dslog`, and `drlog` tools now log empty input lines. The
  `--skip-empty`/`-e` command line option has been added to disable
  logging empty lines.
- The `dmon` tool has gained a `--max-respawns`/`-m` command line option which
  can be used to specify how many times to respawn monitored processes
  before exiting. (Patch by Matt Schulte <<mschulte@wyze.com>>.)
- New `libsetunbuf.so` helper which can be used to disable buffering of
  the standard output stream on arbitrary programs via `LD_PRELOAD`.
  (Patch by Matt Schulte <<mschulte@wyze.com>>.)

### Fixed
- The exit status of monitored processes is now correctly propagated as
  exit code of `dmon` itself.
- The `dlog`, `dslog`, and `drlog` tools will no longer exit unexpectedly
  when they receive an empty input line. (Patch by Matt Schulte
  <<mschulte@wyze.com>>.)

## [v0.5.0] - 2020-08-27
### Added
- Support listing applets compiled into a multicall `dmon` binary when the
  `DMON_LIST_MULTICALL_APPLETS` environment variable is set and non-zero.

### Changed
- GNU Make is not required anymore, the included `Makefile` now works with
  the BSD variant as well, and probably others.

## [v0.4.5] - 2018-08-05
### Fixed
- Make it possible to build `dmon` with newer GCC versions.

## [v0.4.4] - 2016-10-30
### Fixed
- Correctly forward signals to the log process when using `-S`/`--log-sigs`.

## [v0.4.3] - 2016-10-13
### Added
- Allow setting the work directory with `-W`/`--work-dir` in the command line
  (or `work-dir` in a configuration file).

## [v0.4.2] - 2012-01-07
### Added
- Manual pages now also include the long command line options.
- `dlog` can now be instructed to prefix every log message with a given string,
  using the `-p`/`--prefix` command line option.
- `dlog`, `drlog`, and `dslog` have gained the `-i`/`--input-fd` command line
  option, which allows reading log messages from file descriptors different from
  the standard input.

### Fixed
- `dlog` and `drlog` now handle the `INT`, `TERM`, and `HUP` signals gracefully.

[Unreleased]: https://github.com/aperezdc/dmon/compare/v0.5.1...HEAD
[v0.5.1]: https://github.com/aperezdc/dmon/compare/v0.5.0...v0.5.1
[v0.5.0]: https://github.com/aperezdc/dmon/compare/v0.4.5...v0.5.0
[v0.4.5]: https://github.com/aperezdc/dmon/compare/v0.4.4...v0.4.5
[v0.4.4]: https://github.com/aperezdc/dmon/compare/v0.4.3...v0.4.4
[v0.4.3]: https://github.com/aperezdc/dmon/compare/v0.4.2...v0.4.3
[v0.4.2]: https://github.com/aperezdc/dmon/compare/v0.4.1...v0.4.2


dslog.c
/*
 * dslog.c
 * Copyright (C) 2010-2020 Adrian Perez <aperez@igalia.com>
 *
 * Distributed under terms of the MIT license.
 */

#include "deps/cflag/cflag.h"
#include "deps/dbuf/dbuf.h"
#include "util.h"
#include <assert.h>
#include <stdlib.h>
#include <syslog.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>


#if !(defined(MULTICALL) && MULTICALL)
# define dslog_main main
#endif /* MULTICALL */


#ifndef DEFAULT_FACILITY
#define DEFAULT_FACILITY "daemon"
#endif /* !DEFAULT_FACILITY */


#ifndef DEFAULT_PRIORITY
#define DEFAULT_PRIORITY "warning"
#endif /* !DEFAULT_PRIORITY */


static int
name_to_facility (const char *name)
{
    static const struct {
        const char *name;
        int         code;
    } map[] = {
#ifdef LOG_AUTHPRIV
        { "auth",     LOG_AUTHPRIV },
#else  /* LOG_AUTHPRIV */
        { "auth",     LOG_AUTH     },
#endif /* LOG_AUTHPRIV */
        { "cron",     LOG_CRON     },
        { "daemon",   LOG_DAEMON   },
        { "ftp",      LOG_FTP      },
        { "kern",     LOG_KERN     },
        { "kernel",   LOG_KERN     },
        { "local0",   LOG_LOCAL0   },
        { "local1",   LOG_LOCAL1   },
        { "local2",   LOG_LOCAL2   },
        { "local3",   LOG_LOCAL3   },
        { "local4",   LOG_LOCAL4   },
        { "local5",   LOG_LOCAL5   },
        { "local6",   LOG_LOCAL6   },
        { "local7",   LOG_LOCAL7   },
        { "lpr",      LOG_LPR      },
        { "print",    LOG_LPR      },
        { "printer",  LOG_LPR      },
        { "mail",     LOG_MAIL     },
        { "news",     LOG_NEWS     },
        { "user",     LOG_USER     },
        { "uucp",     LOG_UUCP     },
        { NULL,       0            },
    };
    unsigned i = 0;

    while (map[i].name) {
        if (strcasecmp (name, map[i].name) == 0)
            return map[i].code;
        i++;
    }

    return -1;
}


static int
name_to_priority (const char *name)
{
    static const struct {
        const char *name;
        int         code;
    } map[] = {
        { "emerg",     LOG_EMERG   },
        { "emergency", LOG_EMERG   },
        { "alert",     LOG_ALERT   },
        { "crit",      LOG_CRIT    },
        { "critical",  LOG_CRIT    },
        { "err",       LOG_ERR     },
        { "error",     LOG_ERR     },
        { "warn",      LOG_WARNING },
        { "warning",   LOG_WARNING },
        { "notice",    LOG_NOTICE  },
        { "info",      LOG_INFO    },
        { "debug",     LOG_DEBUG   },
        { NULL,        0           },
    };
    unsigned i = 0;

    while (map[i].name) {
        if (strcasecmp (map[i].name, name) == 0)
            return map[i].code;
        i++;
    }

    return -1;
}


static bool running = true;


static enum cflag_status
_facility_option(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    int facility = name_to_facility(arg);
    if (facility == -1)
        return CFLAG_BAD_FORMAT;

    *((int*) spec->data) = facility;
    return CFLAG_OK;
}


static enum cflag_status
_priority_option(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    int priority = name_to_priority(arg);
    if (priority == -1)
        return CFLAG_BAD_FORMAT;

    *((int*) spec->data) = priority;
    return CFLAG_OK;
}


int
dslog_main (int argc, char **argv)
{
    int flags = 0;
    int in_fd = STDIN_FILENO;
    int facility = name_to_facility (DEFAULT_FACILITY);
    int priority = name_to_priority (DEFAULT_PRIORITY);
    bool console = false;
    bool skip_empty = false;
    char *env_opts = NULL;
    struct dbuf linebuf = DBUF_INIT;
    struct dbuf overflow = DBUF_INIT;

    struct cflag dslog_options[] = {
        {
            .name = "facility", .letter = 'f',
            .help = "Log facility (default: daemon).",
            .func = _facility_option,
            .data = &facility,
        },
        {
            .name = "priority", .letter = 'p',
            .help = "Log priority (default: warning).",
            .func = _priority_option,
            .data = &priority,
        },
        CFLAG(int, "input-fd", 'i', &in_fd,
              "File descriptor to read input from (default: stdin)."),
        CFLAG(bool, "console", 'c', &console,
              "Log to console if sending messages to logger fails."),
        CFLAG(bool, "skip-empty", 'e', &skip_empty,
              "Ignore empty lines with no characters."),
        CFLAG_HELP,
        CFLAG_END
    };

    if ((env_opts = getenv ("DSLOG_OPTIONS")) != NULL)
        replace_args_string (env_opts, &argc, &argv);

    const char *argv0 = cflag_apply(dslog_options, "[options] name", &argc, &argv);

    if (console)
        flags |= LOG_CONS;

    /* We will be no longer using standard output. */
    close (STDOUT_FILENO);

    if (!argc)
        die ("%s: process name not specified.\n", argv0);

    openlog (argv[0], flags, facility);

    while (running) {
        ssize_t bytes = freadline (in_fd, &linebuf, &overflow, 0);
        if (bytes == 0)
            break; /* EOF */

        if (bytes < 0) {
            fprintf (stderr, "%s: error reading input: %s\n", argv0, ERRSTR);
            exit (111);
        }

        if (!skip_empty || dbuf_size(&linebuf) > 1)
            syslog(priority, "%s", dbuf_str(&linebuf));

        dbuf_clear(&linebuf);
    }

    closelog ();

	exit (EXIT_SUCCESS);
}

/* vim: expandtab shiftwidth=4 tabstop=4
 */
